Assets/Scripts/MazeGenerator.cs

```diff
@@ -0,0 +1,164 @@
+﻿using System.Collections;
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class MazeGenerator : MonoBehaviour
+{
+    [SerializeField]
+    private int width;
+
+    [SerializeField]
+    private int height;
+
+    public Transform brick;
+    private int[,] Maze;
+
+    private enum TileType { path, wall };
+    private List<Vector3> pathMazes = new List<Vector3>();
+    private Stack<Vector2> tileStack = new Stack<Vector2>();
+
+    private readonly List<Vector2> offsets = new List<Vector2>
+    {
+        new Vector2(0, 1),
+        new Vector2(0, -1),
+        new Vector2(1, 0),
+        new Vector2(-1, 0)
+    };
+
+    private System.Random rnd = new System.Random();
+
+    private Vector2 currentTile;
+    public Vector2 CurrentTile
+    {
+        get { return currentTile; }
+
+        private set
+        {
+            if (value.x < 1 || value.x >= this.width - 1 || value.y < 1 || value.y >= this.height - 1)
+            {
+                throw new ArgumentException("CurrentTile must be within the one tile border all around the maze");
+            }
+            else if (value.x % 2 == 1 || value.y % 2 == 1)
+            {
+                currentTile = value;
+            }
+            else
+            {
+                throw new ArgumentException("The current square must not be both on an even X-axis and an even Y-axis, to ensure we can get walls around all tunnels");
+            }
+        }
+    }
+
+    private static MazeGenerator instance;
+    public static MazeGenerator Instance
+    {
+        get { return instance; }
+    }
+
+    private void Awake()
+    {
+        instance = this;
+        GenerateMaze();
+    }
+
+
+    private void GenerateMaze()
```

Comment:
could we make something a bit more generic such that we could swap maze generation algorithms? For example: in the future a designer could say that they want the players to decide the algorithm used. You could then implement Kruskals algorithm and make that be able to swap in? You don't have to implement this, but I do want you to go over this with Trevor or me